### DISCUSSION TITLE
Using SWITCH statement in formula adds an @ in Excel

### DISCUSSION THREAD

**Discussion by jhorzek** [JanMarvin/openxlsx2#1192] [Q&A] [answered]
Hi,

Thank you very much for the great package! I am currently trying to implement some automations and ran into a an issue when using SWITCH in a formula within a data table. Here is a minimal example:

```
library(openxlsx2)

wb <- wb_workbook()
example_data <-
  data.frame(
    sales_price = c(20, 30, 40),
    COGS = c(5, 11, 13),
    sales_quantity = c(1, 2, 3)
  )

## base R method
example_data$switch1       <- "=SWITCH(daily_sales[[#This Row],[sales_price]], 20, 1, 0)"
example_data$switch2       <- "=IF(daily_sales[[#This Row],[sales_price]] > 0, SWITCH(daily_sales[[#This Row],[sales_price]], 20, 1, 0), NA())"
class(example_data$switch1)       <- c(class(example_data$switch1),       "formula")
class(example_data$switch2)       <- c(class(example_data$switch2),       "formula")
wb$
  add_worksheet("Daily Sales")$
  add_data_table(
    x           = example_data,
    table_style = "TableStyleMedium2",
    table_name  = "daily_sales",
  )
wb$save(file = "switch.xlsx",
        overwrite = TRUE)
```

When opening the file in Excel, I get #NAME? errors. This is due to an @ that is added at the beginning of the SWITCH statement. That is, the formula in Excel is `=@SWITCH([@[sales_price]], 20, 1, NA())` instead of `=SWITCH([@[sales_price]], 20, 1, NA())`. When using the SWITCH in an IF statement, the @ gets pushed in front of the IF, which also results in an error.

I am using R 4.4.0 on Windows 11 with openxlsx2 version 1.11.0.9000.

Best,
Jannik

**Comment by JanMarvin** [comment, marked as answer]
Hi @jhorzek , thank you for your kind words!

The error is related to the following:

As can be seen the [list of Excel functions](https://support.microsoft.com/en-us/office/excel-functions-alphabetical-b3944572-255d-4efb-bb96-c6d90033e188), `SWITCH()` is a modern Excel formula (the label MS365 gives the hint)
<img width="816" alt="Screenshot 2024-11-29 at 17 03 11" src="https://github.com/user-attachments/assets/d1efaa3a-965a-4b70-9117-01da0aaefffd">

To use this function you have to add the prefix `_xlfn.` (see [here](https://janmarvin.github.io/ox2-book/chapters/openxlsx2_formulas_manual.html#modern-spreadsheet-functions)). There is another nasty thing:
The function is a cell metadata function - basically an array function with a special tag to hide that it is an array function. You can use the code below, which will make the function work as expected, but will show its array roots.

I haven't really researched cm-functions, it is possible to write them with `wb_add_formula(cm = TRUE)` but they print a warning since their initial implementation and unfortunately it is not yet possible to add them via the `class()`-approach you want to use in the code below.

``` r
library(openxlsx2)

wb <- wb_workbook()
example_data <-
  data.frame(
    sales_price = c(20, 30, 40),
    COGS = c(5, 11, 13),
    sales_quantity = c(1, 2, 3)
  )

## base R method
example_data$switch1       <- "=_xlfn.SWITCH(daily_sales[[#This Row],[sales_price]], 20, 1, 0)"
example_data$switch2       <- "=IF(daily_sales[[#This Row],[sales_price]] > 0, _xlfn.SWITCH(daily_sales[[#This Row],[sales_price]], 20, 1, 0), NA())"
class(example_data$switch1)       <- c(class(example_data$switch1),       "array_formula")
class(example_data$switch2)       <- c(class(example_data$switch2),       "array_formula")
wb$
  add_worksheet("Daily Sales")$
  add_data_table(
    x           = example_data,
    table_style = "TableStyleMedium2",
    table_name  = "daily_sales",
  )
if (interactive()) wb$open()
```

PS: I move this question to discussion, because it is working as expected. The only way to avoid this, would be to always try to keep up with the Excel formula list and parse the formulas ... something I do not really desire.

**Reply by JanMarvin** [reply]
We create the cell metadata information here:

https://github.com/JanMarvin/openxlsx2/blob/0b247cd421f14eecae1049a0cee5b73952e5e6ef/R/write.R#L1240-L1270

But as said, it worked for a few examples, but haven't looked into it further (and ... it's not yet back on the agenda 😃 ).

**Reply by jhorzek** [reply]
Oh wow, thank you very much for this super swift response! The _xlfn prefix solved the issue perfectly :)
